src/wasm/wasm-type.cpp

```diff
@@ -284,6 +284,60 @@ struct FiniteShapeEquator {
   bool eq(const Rtt& a, const Rtt& b);
 };
 
+// A wrapper around a RecGroup that provides equality and hashing based on the
+// structure of the group such that isorecursively equivalent recursion groups
+// will compare equal and will have the same hash. Assumes that all recursion
+// groups reachable from this one have been canonicalized, except for the
+// wrapped group itself.
+struct RecGroupStructure {
+  RecGroup group;
+
+  bool operator==(const RecGroupStructure& other) const;
+  size_t hash() const;
+
+private:
+  bool topLevelEq(HeapType a, HeapType b, RecGroup other) const;
```

Comment:
How should I interpret "other" in all of these? How does it relate to a, b, and the group in this wrapper?

Reply:
`other` needs to be piped down through all of these intermediate functions to reach the leaf `eq` function that compares `HeapType`s without further recursion. That function uses it to detect self-references in that other rec group so they can compare equal to self-references in the wrapped rec group.

Reply:
Sorry, but I must be still missing something obvious here. The name "topLevelEq" suggests to me that it compares a and b at the top level. But what is "other" here? The group of a? Or of b? Is a in this group, and b in other, maybe?

Reply:
Ultimately we are comparing the wrapped rec group and the `other` rec group. In all the `eq` helper functions, `a` comes from the wrapped group and `b` comes from `other`.

Reply:
I see, thanks. Maybe mentioning that in a comment would be helpful.

Reply:
Or, `bGroup` might clarify that it is the group of b? not sure that's great though.